contracts/0.4.24/nos/NodeOperatorsRegistry.sol

```diff
@@ -699,4 +787,158 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, IsContract, AragonApp
+    function getVersion() external view returns (uint256) {
+        return CONTRACT_VERSION_POSITION.getStorageUint256();
+    }
+
+    /**
+     * @notice A function to finalize upgrade to v2 (from v1). Can be called only once
+     * @dev Value 1 in CONTRACT_VERSION_POSITION is skipped due to change in numbering
+     * For more details see https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-10.md
+     */
+    function finalizeUpgrade_v2(address _steth, bytes32 _type) external {
+        require(CONTRACT_VERSION_POSITION.getStorageUint256() == 0, "WRONG_BASE_VERSION");
+
+        _initialize_v2(_steth, _type);
+    }
+
+    function _initialize_v2(address _steth, bytes32 _type) internal {
+        require(_steth != address(0), "STETH_ADDRESS_ZERO");
+
+        STETH_POSITION.setStorageAddress(_steth);
+        TYPE_POSITION.setStorageBytes32(_type);
+
+        uint256 totalOperators = getNodeOperatorsCount();
+        uint256 totalUsedKeys;
+        uint256 totalStoppedKeys;
+        for (uint256 operatorId = 0; operatorId < totalOperators; ++operatorId) {
+            NodeOperator memory operator = operators[operatorId];
+            if (!operator.active) {
+                _trimUnusedNodeOperatorKeys(operatorId);
+            }
+            totalUsedKeys += operator.usedSigningKeys;
+            totalStoppedKeys += operator.stoppedValidators;
+        }
+
+        _setTotalActiveKeys(to64(totalUsedKeys) - to64(totalStoppedKeys));
+        _updateTotalAvailableKeysCount();
+        _increaseKeysOpIndex();
+
+        CONTRACT_VERSION_POSITION.setStorageUint256(2);
+        emit ContractVersionSet(2);
+        emit StethContractSet(_steth);
+        emit SetStakingModuleType(_type);
+    }
+
+    function getType() external view returns (bytes32) {
+        return TYPE_POSITION.getStorageBytes32();
+    }
+
+    function getNodeOperatorActiveKeysCount(uint256 _operatorId) external view operatorExists(_operatorId) returns (uint256) {
+        NodeOperator storage operator = operators[_operatorId];
+        return operator.usedSigningKeys - operator.stoppedValidators;
+    }
+
+    function getNodeOperatorAvailableKeysCount(uint256 _operatorId) external view operatorExists(_operatorId) returns (uint256) {
+        return _getNodeOperatorAvailableKeysCount(_operatorId);
+    }
+
+    function getActiveKeysCount() external view returns (uint256) {
+        return keysUsageStats.totalActiveKeys;
+    }
+
+    function getAvailableKeysCount() external view returns (uint256) {
+        return keysUsageStats.totalAvailableKeys;
+    }
+
+    function getKeysUsageData() external view returns (uint256 activeKeysCount, uint256 availableKeysCount) {
+        KeysUsageStats memory _keysUsageStats = keysUsageStats;
+        activeKeysCount = _keysUsageStats.totalActiveKeys;
+        availableKeysCount = _keysUsageStats.totalAvailableKeys;
+    }
+
+    function distributeRewards() external returns (uint256 distributed) {
```

Comment:
have a corner case if rewards weren't distributed before the last completed oracle report (can change shares)
want to have the claimable model (every NO claims rewards manually)

Reply:
The possible workaround is adding as an Oracle callback

Reply:
Though, better to implement more sane model for withdrawals